### PR TITLE
Implementing python's `global`

### PR DESCRIPTION
**Pull request by oxisto**
This PR tries to implement python's `global` keyword and tries to replicate python variable resolution as good as possible.

### CODE REVIEW

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/scopes/Scope.kt [hunk 1]
```diff
@@ -90,6 +90,9 @@ abstract class Scope(
      */
     @Transient var wildcardImports: MutableSet<ImportDeclaration> = mutableSetOf()
 
+    /** A list of scope modifications. */
+    @Transient var scopeModifications: MutableMap<Name, Scope> = mutableMapOf()
```
Comment (oxisto) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/scopes/Scope.kt:94]: this would be the alternative approach to the node. Depending on what we decide, this would be removed

**Review by maximiliankaul — COMMENTED** [review]
I'd appreciate to have two PRs:
- one for the CPG changes (with it's own tests, so that we can see how it's supposed to work in general and not only in the specific python use-case)
- one making use of the first PR by implementing the Python stuff

I get why you modify the scope of a `Reference` as this is how the Python front-end works. However, don't we need a scope modifier for the declaration (I get that this will be problematic with the Python "add declarations pass"?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/StatementBuilder.kt [hunk 2]
```diff
@@ -329,3 +330,22 @@ fun MetadataProvider.newDefaultStatement(rawNode: Any? = null): DefaultStatement
     log(node)
     return node
 }
+
+/**
+ * Creates a new [DefaultStatement]. The [MetadataProvider] receiver will be used to fill different
```
Comment (maximiliankaul) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/StatementBuilder.kt:335]: ReferenceScopeModifierStatement

File: cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/statementHandler/StatementHandlerTest.kt [hunk 3]
```diff
@@ -232,4 +232,70 @@ class StatementHandlerTest : BaseTest() {
...
+
+    @Test
+    fun testGlobal() {
+        var file = topLevel.resolve("global.py").toFile()
+        val result = analyze(listOf(file), topLevel, true) { it.registerLanguage<PythonLanguage>() }
+        assertNotNull(result)
+
+        // There should be three variable declarations, two local and one global
+        var cVariables = result.variables("c")
+        assertEquals(3, cVariables.size)
```
Comment (maximiliankaul) [inline cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/statementHandler/StatementHandlerTest.kt:244]: I see the global one. But where is the second local one? I only see the `c = 3`

Comment (maximiliankaul) [reply cpg-language-python/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/python/statementHandler/StatementHandlerTest.kt:244]: Ah, the error one.

**Comment by oxisto** [comment]
> I'd appreciate to have two PRs:
> 
> * one for the CPG changes (with it's own tests, so that we can see how it's supposed to work in general and not only in the specific python use-case)
> * one making use of the first PR by implementing the Python stuff

Done. https://github.com/Fraunhofer-AISEC/cpg/pull/1742

> 
> I get why you modify the scope of a `Reference` as this is how the Python front-end works. However, don't we need a scope modifier for the declaration (I get that this will be problematic with the Python "add declarations pass"?

The declaration's scope is not modified. Only the "lookup scope" of the symbol that you are trying to resolve in the current scope is modified (e.g., from the local to the global scope).

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt [hunk 4]
```diff
@@ -537,6 +527,35 @@ class StatementHandler(frontend: PythonLanguageFrontend) :
         return wrapDeclarationToStatement(result)
     }
 
+    private fun handleGlobal(global: Python.AST.Global): LookupScopeStatement {
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:530]: Please add a doc string similar to the other functions.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt [hunk 5]
```diff
@@ -537,6 +527,35 @@ class StatementHandler(frontend: PythonLanguageFrontend) :
...
+            frontend.scopeManager.globalScope?.children?.firstOrNull { it is NameScope }
+
+        return newLookupScopeStatement(
+            global.names.map { parseName(it).localName },
+            pythonGlobalScope,
+            rawNode = global
+        )
+    }
+
+    private fun handleNonLocal(global: Python.AST.Nonlocal): LookupScopeStatement {
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:544]: Please add a doc string similar to the other functions.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt [hunk 6]
```diff
@@ -537,6 +527,35 @@ class StatementHandler(frontend: PythonLanguageFrontend) :
...
+        // We need to find the first outer function scope, or rather the block scope belonging to
+        // the function
+        var outerFunctionScope =
+            frontend.scopeManager.firstScopeOrNull {
+                it is BlockScope && it != frontend.scopeManager.currentScope
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:549]: Do we need to consider nested functions here, too?

Comment (oxisto) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:549]: Unfortunately, I cannot test this, as we are not parsing nested function at all at the moment (see #1736), but I am beginning to hate python even more. It seems that the nonlocal statement actually jumps to the nested function scope where the variable is "declared" and not just the next one. See example
```
def level1():
    c = 1
    def level2():
        d = 1
        def level3():
            nonlocal c, d
            c = c + 2
            d = d + 1
            print(c)
        
        level3()
    level2()
level1()
```

In this case the nonlocal c "jumps" to level 1 and nonlocal d jumps to level 2, but we cannot retrieve this information in the frontend so I would take this as a close approximation and ignore this very narrow corner case.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt [hunk 7]
```diff
@@ -79,59 +78,105 @@ class PythonAddDeclarationsPass(ctx: TranslationContext) : ComponentPass(ctx) {
         }
     }
 
-    /*
-     * Return null when not creating a new decl
+    /**
+     * This function will create a new dynamic [VariableDeclaration] if there is a write access to
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:82]: This function does not check whether there is a write access. However, the doc implies this. Maybe rephrase to "should be called when handling write access invocations"?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt [hunk 8]
```diff
@@ -79,59 +78,105 @@ class PythonAddDeclarationsPass(ctx: TranslationContext) : ComponentPass(ctx) {
...
+
+        // First, check if we need to create a field
+        var field: FieldDeclaration? =
+            when {
+                // Check, whether we are referring to a "self.X", which would create a field
+                scopeManager.isInRecord && scopeManager.isInFunction && ref.refersToReceiver -> {
+                    // We need to temporarily jump into the scope of the current record to
+                    // add the field. These are instance attributes
+                    scopeManager.withScope(scopeManager.firstScopeIsInstanceOrNull<RecordScope>()) {
+                        newFieldDeclaration(ref.name)
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:119]: Copy code and location from the reference?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt [hunk 9]
```diff
@@ -79,59 +78,105 @@ class PythonAddDeclarationsPass(ctx: TranslationContext) : ComponentPass(ctx) {
...
                 }
+                scopeManager.isInRecord && scopeManager.isInFunction && ref is MemberExpression -> {
+                    // If this is any other member expression, we are usually not interested in
+                    // creating fields, except if this is a receiver
+                    return null
+                }
+                scopeManager.isInRecord -> {
+                    // We end up here for fields declared directly in the class body. These are
+                    // class attributes; more or less static fields.
+                    newFieldDeclaration(ref.name)
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:130]: Copy code and location from the reference?

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt [hunk 10]
```diff
@@ -79,59 +78,105 @@ class PythonAddDeclarationsPass(ctx: TranslationContext) : ComponentPass(ctx) {
...
+                }
+            }
+
+        // If we didn't create any field up to this point and if we are still have not returned, we
+        // can create a normal variable. We need to take scope modifications into account.
+        var decl =
+            if (field == null && targetScope != null) {
+                scopeManager.withScope(targetScope) { newVariableDeclaration(ref.name) }
+            } else if (field == null) {
+                newVariableDeclaration(ref.name)
```
Comment (maximiliankaul) [inline cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:143]: Copy code and location from the reference?

**Review by maximiliankaul — APPROVED** [review]

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt (hunk 7 above)
Comment (oxisto) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:82]: I was actually thinking in adding another check so I can do that, should be minimal effort for the compiler.

Comment (oxisto) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:82]: Then I can also get rid of the weird check if this is part of a call expression

Comment (oxisto) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:82]: There is a bug with write references in a named expression. https://github.com/Fraunhofer-AISEC/cpg/pull/1785 needs to be resolved first.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt (hunk 6 above)
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt:557]: Will be addressed in #1736 -> resolved for now.

File: cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt (hunk 7 above)
Comment (maximiliankaul) [reply cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/PythonAddDeclarationsPass.kt:81]: Check implemented
